### PR TITLE
Use throw helpers in a few more places

### PR DESCRIPTION
**Pull request by stephentoub**

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/String.cs [hunk 1]
```diff
@@ -295,13 +295,19 @@ private static unsafe string Ctor(ReadOnlySpan<char> value)
 
         public static string Create<TState>(int length, TState state, SpanAction<char, TState> action)
         {
-            ArgumentNullException.ThrowIfNull(action);
+            if (action is null)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.action);
+            }
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/String.cs:301]: Why this change, particularly given it changes the exception type?

**Review by tannergooding — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/String.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/String.cs:301]: Changing the exception type is a typo; that I need to fix.  This pattern results in ever so slightly smaller asm, and as it's in a generic method and we already had the helper and the appropriately named constant argument, it was trivial to switch.

File: src/libraries/System.Private.CoreLib/src/System/String.cs [hunk 2]
```diff
@@ -295,13 +295,19 @@ private static unsafe string Ctor(ReadOnlySpan<char> value)
 
         public static string Create<TState>(int length, TState state, SpanAction<char, TState> action)
         {
-            ArgumentNullException.ThrowIfNull(action);
+            if (action is null)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.action);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/String.cs:300]:
```suggestion
                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.action);
```

File: src/libraries/System.Private.CoreLib/src/System/String.cs (hunk 1 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/String.cs]: Probably worth a small comment in code so this isn't undone by some future refactoring without validating the change

File: src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/BrotliStream.cs [hunk 3]
```diff
@@ -61,8 +61,7 @@ public BrotliStream(Stream stream, CompressionMode mode, bool leaveOpen)
 
         private void EnsureNotDisposed()
```
Comment (xtqqczze) [inline src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/BrotliStream.cs:62]:
```suggestion
        [MemberNotNull(nameof(_stream))]
        private void EnsureNotDisposed()
```

File: src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateManagedStream.cs [hunk 4]
```diff
@@ -148,13 +148,7 @@ public override int ReadByte()
 
         private void EnsureNotDisposed()
```
Comment (xtqqczze) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateManagedStream.cs:149]:
```suggestion
        [MemberNotNull(nameof(_stream))]
        private void EnsureNotDisposed()
```

File: src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs [hunk 5]
```diff
@@ -326,11 +326,7 @@ internal int ReadCore(Span<byte> buffer)
 
         private void EnsureNotDisposed()
```
Comment (xtqqczze) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs:327]:
```suggestion
        [MemberNotNull(nameof(_stream))]
        private void EnsureNotDisposed()
```

File: src/libraries/System.IO.Compression/src/System/IO/Compression/GZipStream.cs [hunk 6]
```diff
@@ -242,15 +242,7 @@ public override Task CopyToAsync(Stream destination, int bufferSize, Cancellatio
 
         private void CheckDeflateStream()
```
Comment (xtqqczze) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/GZipStream.cs:243]:
```suggestion
        [MemberNotNull(nameof(_deflateStream))]
        private void CheckDeflateStream()
```

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZLibStream.cs [hunk 7]
```diff
@@ -279,15 +279,7 @@ public override ValueTask DisposeAsync()
         /// <summary>Throws an <see cref="ObjectDisposedException"/> if the stream is closed.</summary>
         private void ThrowIfClosed()
```
Comment (xtqqczze) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZLibStream.cs:280]:
```suggestion
        [MemberNotNull(nameof(_deflateStream))]
        private void ThrowIfClosed()
```

File: src/libraries/System.Private.CoreLib/src/System/String.cs [hunk 8]
```diff
@@ -295,13 +295,19 @@ private static unsafe string Ctor(ReadOnlySpan<char> value)
 
         public static string Create<TState>(int length, TState state, SpanAction<char, TState> action)
```
Comment (xtqqczze) [inline src/libraries/System.Private.CoreLib/src/System/String.cs:296]:
```suggestion
        public static string Create<TState>(int length, TState state, [NotNull] SpanAction<char, TState> action)
```

File: src/mono/System.Private.CoreLib/src/System/GC.Mono.cs [hunk 9]
```diff
@@ -116,8 +116,7 @@ public static void KeepAlive(object? obj)
         public static int GetGeneration(WeakReference wo)
         {
             object? obj = wo.Target;
-            if (obj == null)
-                throw new ArgumentException(null, nameof(wo));
+            ArgumentNullException.ThrowIfNull(obj, nameof(wo));
```
Comment (xtqqczze) [inline src/mono/System.Private.CoreLib/src/System/GC.Mono.cs:119]: Is use of `ArgumentNullException` the most appropriate exception, given that the parameter `wo` is not null?

**Review by xtqqczze — COMMENTED** [review]
Don't forgot the nullable annotations :)

File: src/libraries/System.Private.CoreLib/src/System/String.cs (hunk 8 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/String.cs:296]: The parameter is already non-nullable.  Such annotation is duplicative.

File: src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/BrotliStream.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/BrotliStream.cs:62]: These annotations weren't there previously.  These changes don't change the need for them.  The only benefit to such annotation would be if someone needed to know that _stream is non-nullable after the Ensure call, and that doesn't exist today.  Such annotation can be added to internal/private members as needed.

Comment (xtqqczze) [reply src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/BrotliStream.cs:62]: The changes introduce a method call. I believe `MemberNotNullAttribute` may provide a benefit similar to `DoesNotReturnAttribute`, as code analysis can determine the method will not return when the member is null.

**Comment by xtqqczze** [comment]
I believe `MemberNotNullAttribute` may provide a benefit similar to `DoesNotReturnAttribute`, as code analysis can determine the method will not return when the member is null.

**Comment by stephentoub** [comment]
> I believe MemberNotNullAttribute may provide a benefit similar to DoesNotReturnAttribute, as code analysis can determine the method will not return when the member is null.

These methods are private.  If the caller doesn't try to dereference the cited member, such attribution doesn't help with anything.

File: src/mono/System.Private.CoreLib/src/System/GC.Mono.cs (hunk 9 above)
Comment (stephentoub) [reply src/mono/System.Private.CoreLib/src/System/GC.Mono.cs:119]: It's the same exception coreclr already throws for it here:
https://github.com/dotnet/runtime/blob/4f63db1b9fbee35e3749ab7deb949cd8de63e848/src/coreclr/vm/comutilnative.cpp#L826-L827

File: src/libraries/System.Private.CoreLib/src/System/String.cs [hunk 10]
```diff
@@ -295,13 +295,19 @@ private static unsafe string Ctor(ReadOnlySpan<char> value)
...
         {
-            ArgumentNullException.ThrowIfNull(action);
+            if (action is null)
```
Comment (drieseng) [inline src/libraries/System.Private.CoreLib/src/System/String.cs:298]: @stephentoub: Just out of curiosity, why was this changed?

Not related to this PR: in `Ctor(char c, int count)` the count is checked twice:
https://github.com/stephentoub/runtime/blob/3298c130f2540f252647cfe0129ed93fb0affcf1/src/libraries/System.Private.CoreLib/src/System/String.cs#L268-L272

Better to use `ThrowHelper.ThrowArgumentOutOfRangeException(...)` there?

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/String.cs:298]: This pattern results in ever so slightly smaller asm, and as it's in a generic method and we already had the helper and the appropriately named constant argument, it was trivial to switch.
